Mathlib/SetTheory/ZFC/VonNeumann.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.SetTheory.ZFC.Class
+import Mathlib.SetTheory.ZFC.Ordinal
+import Mathlib.SetTheory.ZFC.Rank
+
+/-!
+# Von Neumann hierarchy
+
+This file defines the von Neumann hierarchy of sets `V_ o` for ordinal `o`, which is recursively
+defined so that `V_ a = ⋃ b < a, powerset (V_ b)`. This stratifies the universal class, in the sense
+that `⋃ o, V_ o = univ`.
+
+## Notation
+
+- `V_ o` is notation for `vonNeumann o`. It is scoped in the `ZFSet` namespace.
+-/
+
+universe u
+
+open Order
+
+namespace ZFSet
+
+/-- The von Neumann hierarchy is defined so that `V_ o` is the union of the powersets of all
+`V_ a` for `a < o`. It satisfies the following properties:
+
+- `vonNeumann_zero`: `V_ 0 = ∅`
+- `vonNeumann_succ`: `V_ (succ a) = powerset (V_ a)`
+- `vonNeumann_of_isSuccPrelimit`: `IsSuccPrelimit a → V_ a = ⋃ b < a, V_ b`
+-/
+noncomputable def vonNeumann (o : Ordinal.{u}) : ZFSet.{u} :=
+  ⋃₀ range fun a : Set.Iio o ↦ powerset (vonNeumann a)
+termination_by o
+decreasing_by exact a.2
+
+@[inherit_doc]
+scoped notation "V_ " => vonNeumann
+
+variable {a b o : Ordinal.{u}} {x : ZFSet.{u}}
+
+lemma mem_vonNeumann' : x ∈ V_ o ↔ ∃ a < o, x ⊆ V_ a := by rw [vonNeumann]; simp
+
+theorem isTransitive_vonNeumann (o : Ordinal) : IsTransitive (V_ o) := by
+  rw [vonNeumann]
+  refine IsTransitive.sUnion' fun x hx ↦ ?_
+  obtain ⟨⟨a, _⟩, rfl⟩ := mem_range.1 hx
+  exact (isTransitive_vonNeumann a).powerset
+termination_by o
+
+theorem vonNeumann_mem_of_lt (h : a < b) : V_ a ∈ V_ b := by
```

Comment:
Not sure, but maybe these `gcongr` tags would be useful?
```suggestion
@[gcongr] theorem vonNeumann_mem_of_lt (h : a < b) : V_ a ∈ V_ b := by
```

Reply:
Can't do harm.